python/cuml/testing/strategies.py

```diff
@@ -12,12 +12,285 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import cudf
+import cupy as cp
+import numpy as np
 from hypothesis import assume
-from hypothesis.extra.numpy import arrays, floating_dtypes
-from hypothesis.strategies import composite, integers, just, none, one_of
+from hypothesis.extra.numpy import array_shapes, arrays, floating_dtypes
+from hypothesis.strategies import (
+    composite,
+    integers,
+    just,
+    none,
+    one_of,
+    sampled_from,
+)
 from sklearn.datasets import make_regression
 from sklearn.model_selection import train_test_split
 
+from cuml.common.array import CumlArray
+
+_CUML_ARRAY_INPUT_TYPES = ["numpy", "cupy", "series"]
+
+
+_CUML_ARRAY_DTYPES = [
+    np.float16,
+    np.float32,
+    np.float64,
+    np.int8,
+    np.int16,
+    np.int32,
+    np.int64,
+    np.uint8,
+    np.uint16,
+    np.uint32,
+    np.uint64,
+]
+
+_CUML_ARRAY_ORDERS = ["F", "C"]
+
+
+_CUML_ARRAY_OUTPUT_TYPES = [
+    "cudf",
+    "cupy",
+    "dataframe",
+    "numba",
+    "numpy",
+    "series",
+]
+
+
+UNSUPPORTED_CUDF_DTYPES = [
+    np.uint8,
+    np.uint16,
+    np.uint32,
+    np.uint64,
+    np.float16,
+]
+
+
+@composite
+def cuml_array_input_types(draw):
+    """Generates all supported cuml array input types."""
+    return draw(sampled_from(_CUML_ARRAY_INPUT_TYPES))
+
+
+@composite
+def cuml_array_output_types(draw):
+    """Generates all cuml array supported output types."""
+    return draw(sampled_from(_CUML_ARRAY_OUTPUT_TYPES))
+
+
+@composite
+def cuml_array_output_dtypes(draw):
+    """Generates all cuml array supported output dtypes."""
+    return draw(sampled_from(_CUML_ARRAY_DTYPES))
+
+
+@composite
+def cuml_array_dtypes(draw):
+    """Generates all supported cuml array dtypes."""
+    return draw(sampled_from(_CUML_ARRAY_DTYPES))
+
+
+@composite
+def cuml_array_orders(draw):
+    """Generates all supported cuml array orders."""
+    return draw(sampled_from(_CUML_ARRAY_ORDERS))
+
+
+@composite
+def cuml_array_shapes(
+    draw, *, min_dims=1, max_dims=2, min_side=1, max_side=None
+):
+    """
+    Generates cuml array shapes.
+
+    See also: hypothesis.extra.numpy.array_shapes()
+
+    Parameters
+    ----------
+    min_dims: int, default=1
+        Returned shapes will have at least this number of dimensions.
+    max_dims: int, default=2
+        Returned shapes will have at most this number of dimensions.
+    min_side: int, default=1
```

Comment:
Could we replace `min_side` and `max_size` with `min_extent` and `max_extent`? That would be more consistent with terminology used in RAFT and elsewhere.

Reply:
I'd prefer the RAFT terminology, but I was matching hypothesis' own [`extra.numpy.array_shapes()`](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.array_shapes) API. Does that change your recommendation?

Reply:
It certainly does! Carry on! ;)